src/v/archival/archival_policy.cc

```diff
@@ -96,36 +129,231 @@ archival_policy::lookup_result archival_policy::find_segment(
+    bool force_upload = upload_deadline_reached();
+    if (!closed && !force_upload) {
+        // Fast path, next upload candidate is not yet ready. We may want to
         // optimize this case because it's expected to happen pretty often. This
         // can be done by saving weak_ptr to the segment inside the policy
         // object. The segment must be changed to derive from
         // ss::weakly_referencable.
-        vlog(
-          archival_log.trace,
-          "Upload policy for {}, upload candidate is not closed",
-          _ntp);
+        if (archival_log.is_enabled(ss::log_level::debug)) {
+            vlog(
+              archival_log.debug,
+              "Upload policy for {}, upload candidate is not closed",
+              _ntp);
+        }
         return {};
     }
     auto dirty_offset = (*it)->offsets().dirty_offset;
-    if (dirty_offset > high_watermark) {
+    if (dirty_offset > high_watermark && !force_upload) {
         vlog(
           archival_log.debug,
-          "Upload policy for {}, upload candidate offset {} is below high "
+          "Upload policy for {}, upload candidate offset {} is above high "
           "watermark {}",
+          _ntp,
           dirty_offset,
-          high_watermark,
-          _ntp);
+          high_watermark);
         return {};
     }
-    return {.segment = *it, .ntp_conf = &ntp_conf};
+    if (_upload_limit) {
+        _upload_deadline = ss::lowres_clock::now() + _upload_limit.value()();
+    }
+    return {.segment = *it, .ntp_conf = &ntp_conf, .forced = force_upload};
+}
+
+// Data sink for noop output_stream instance
+// needed to implement scanning
+struct null_data_sink final : ss::data_sink_impl {
+    ss::future<> put(ss::net::packet data) final { return put(data.release()); }
+    ss::future<> put(std::vector<ss::temporary_buffer<char>> all) final {
+        return ss::do_with(
+          std::move(all), [this](std::vector<ss::temporary_buffer<char>>& all) {
+              return ss::do_for_each(
+                all, [this](ss::temporary_buffer<char>& buf) {
+                    return put(std::move(buf));
+                });
+          });
+    }
+    ss::future<> put(ss::temporary_buffer<char>) final { return ss::now(); }
+    ss::future<> flush() final { return ss::now(); }
+    ss::future<> close() final { return ss::now(); }
+};
+
+static ss::output_stream<char> make_null_output_stream() {
+    auto ds = ss::data_sink(std::make_unique<null_data_sink>());
+    ss::output_stream<char> ostr(std::move(ds), 4_KiB);
+    return ostr;
+}
+
+/// This function computes offsets for the upload (inc. file offets)
+/// If the full segment is uploaded the segment is not scanned.
+/// If the upload is partial, the partial scan will be performed if
+/// the segment has the index and full scan otherwise.
+static ss::future<> get_file_range(
+  model::offset off_begin,
+  std::optional<model::offset> off_end,
```

Comment:
better to also note that these are inclusive offsets

Reply:
fixed